File: app/src/main/java/com/example/punchpad2/UndoManager.java
package com.example.punchpad2;

import android.text.Editable;
import android.text.TextWatcher;
import android.widget.EditText;

public class UndoManager {

    private final UndoStack undoStack = new UndoStack();
    private boolean isUserChange = true;
    private EditText editText;

    private final TextWatcher watcher = new TextWatcher() {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            if (isUserChange) {
                undoStack.recordState(s.toString());
            }
        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {}

        @Override
        public void afterTextChanged(Editable s) {}
    };

    public void attach(EditText editText) {
        this.editText = editText;
        undoStack.setInitialState(editText.getText().toString().trim());
        editText.addTextChangedListener(watcher);
    }

    public void clear() {
        undoStack.clear();
    }

    public void undo() {
        if (editText == null || !canUndo()) return;

        isUserChange = false;
        String current = editText.getText().toString();
        String previous = undoStack.undo(current);

        // Strict check prevents clearing to empty by accident
        if (previous != null && !previous.equals(current) && !previous.isEmpty()) {
            editText.setText(previous);
            editText.setSelection(previous.length());
        }

        isUserChange = true;
    }

    public void redo() {
        if (editText == null || !canRedo()) return;

        isUserChange = false;
        String current = editText.getText().toString();
        String next = undoStack.redo(current);

        if (next != null && !next.equals(current)) {
            editText.setText(next);
            editText.setSelection(next.length());
        }

        isUserChange = true;
    }

    public boolean canUndo() {
        return undoStack != null && !undoStack.isEmptyUndo();
    }

    public boolean canRedo() {
        return undoStack != null && !undoStack.isEmptyRedo();
    }
}

File: app/src/main/java/com/stratonotes/UndoManager.kt
package com.stratonotes

import android.text.Editable
import android.text.TextWatcher
import android.widget.EditText
import kotlinx.coroutines.*

class UndoManager(private val editText: EditText) {

    private val undoStack = UndoStack()
    private var lastSnapshot: UndoStack.UndoSnapshot? = null
    private var debounceJob: Job? = null

    init {
        editText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun afterTextChanged(s: Editable?) {}

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                debounceJob?.cancel()
                debounceJob = CoroutineScope(Dispatchers.Main).launch {
                    delay(500)
                    pushSnapshot()
                }
            }
        })
    }

    fun pushSnapshot() {
        val text = editText.text.toString()
        val cursor = editText.selectionStart
        val current = UndoStack.UndoSnapshot(text, cursor)

        if (lastSnapshot != current) {
            undoStack.push(current)
            lastSnapshot = current
        }
    }

    fun undo() {
        val snapshot = undoStack.undo()
        snapshot?.let {
            editText.setText(it.text)
            editText.setSelection(it.cursor.coerceIn(0, it.text.length))
        }
    }

    fun redo() {
        val snapshot = undoStack.redo()
        snapshot?.let {
            editText.setText(it.text)
            editText.setSelection(it.cursor.coerceIn(0, it.text.length))
        }
    }

    fun canUndo(): Boolean = undoStack.canUndo()

    fun canRedo(): Boolean = undoStack.canRedo()

    fun clear() {
        undoStack.clear()
        lastSnapshot = null
    }
}
